core/src/main/java/io/questdb/cutlass/text/TextImportTask.java

```diff
@@ -0,0 +1,995 @@
+                }
+            }
+        }
+
+        private void unmap(FilesFacade ff, DirectLongList mergeIndexes) {
+            for (long i = 0, sz = mergeIndexes.size() / 2; i < sz; i++) {
+                final long addr = mergeIndexes.get(2 * i);
+                final long size = mergeIndexes.get(2 * i + 1) * CsvFileIndexer.INDEX_ENTRY_SIZE;
+                ff.munmap(addr, size, MemoryTag.MMAP_PARALLEL_IMPORT);
+            }
+            mergeIndexes.clear();
+        }
+
+        private boolean onField(long offset, final DirectByteCharSequence dbcs, TableWriter.Row w, int fieldIndex, DirectCharSink utf8Sink)
+                throws TextException {
+            TypeAdapter type = this.types.getQuick(fieldIndex);
+            try {
+                type.write(w, fieldIndex, dbcs, utf8Sink);
+            } catch (Exception ignore) {
+                errors++;
+                logError(offset, fieldIndex, dbcs);
+                switch (atomicity) {
+                    case Atomicity.SKIP_ALL:
+                        tableWriterRef.rollback();
+                        throw TextException.$("bad syntax [line offset=").put(offset).put(",column=").put(fieldIndex).put(']');
+                    case Atomicity.SKIP_ROW:
+                        w.cancel();
+                        return true;
+                    default: // SKIP column
+                        break;
+                }
+            }
+            return false;
+        }
+
+        private void onFieldsPartitioned(final ObjList<DirectByteCharSequence> values, int valuesLength, DirectCharSink utf8Sink) {
+            assert tableWriterRef != null;
+            DirectByteCharSequence dbcs = values.getQuick(timestampIndex);
+            final TableWriter.Row w = getRow(dbcs, offset);
+            if (w == null) {
+                return;
+            }
+            for (int i = 0; i < valuesLength; i++) {
+                dbcs = values.getQuick(i);
+                if (i == timestampIndex || dbcs.length() == 0) {
+                    continue;
+                }
+                if (onField(offset, dbcs, w, i, utf8Sink)) return;
+            }
+            w.append();
+        }
+
+        private TableWriter.Row getRow(DirectByteCharSequence dbcs, long offset) {
+            try {
+                return tableWriterRef.newRow(timestampAdapter.getTimestamp(dbcs));
+            } catch (Exception e) {
+                if (atomicity == Atomicity.SKIP_ALL) {
+                    throw TextException.$("Can't parse timestamp on line starting at offset=").put(offset).put(". ").put(e.getMessage());
+                } else {
+                    logError(offset, timestampIndex, dbcs);
+                    return null;
+                }
+            }
+        }
+
+        private long openIndexChunks(FilesFacade ff, Path partitionPath, DirectLongList mergeIndexes, int partitionLen) {
+            long mergedIndexSize = 0;
+            long chunk = ff.findFirst(partitionPath);
```

Comment:
boilerplate can be replaced with `ff.walk()`

Reply:
ff.walk() requires pushing related state into the object and is even uglier (if we want to avoid allocating closure) .